velox/exec/benchmarks/HashJoinGroupProbeBenchmark.cpp

```diff
@@ -0,0 +1,240 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/HashTable.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+#include <folly/Benchmark.h>
+#include <folly/executors/CPUThreadPoolExecutor.h>
+#include <folly/init/Init.h>
+
+#include <utility>
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::test;
+
+namespace {
+struct HashTableBenchmarkParams {
+  HashTableBenchmarkParams() = default;
+
+  // Benchmark params, we need to provide:
+  //  -the expect hash mode,
+  //  -the build row schema,
+  //  -the expected hash table size,
+  //  -number of building rows,
+  HashTableBenchmarkParams(
+      BaseHashTable::HashMode mode,
+      const TypePtr& buildType,
+      int64_t hashTableSize,
+      int64_t buildSize)
+      : mode{mode},
+        buildType{buildType},
+        hashTableSize{hashTableSize},
+        buildSize{buildSize} {
+    VELOX_CHECK_LE(hashTableSize, buildSize);
+
+    if (hashTableSize > BaseHashTable::kArrayHashMaxSize &&
+        mode == BaseHashTable::HashMode::kArray) {
+      VELOX_FAIL("Bad hash mode.");
+    }
+
+    numFields = buildType->size();
+    if (mode == BaseHashTable::HashMode::kNormalizedKey) {
+      extraValue = BaseHashTable::kArrayHashMaxSize + 100;
+    } else if (mode == BaseHashTable::HashMode::kHash) {
+      extraValue = std::numeric_limits<int64_t>::max() - 1;
+    } else {
+      extraValue = 0;
+    }
+
+    title = fmt::format(
+        "{},Size:{},withDup:{}",
+        BaseHashTable::modeString(mode),
+        buildSize,
+        buildSize > hashTableSize);
+  }
+
+  // Expected mode.
+  BaseHashTable::HashMode mode;
+
+  // Type of build & probe row.
+  TypePtr buildType;
+
+  // Distinct rows in the table.
+  int64_t hashTableSize{};
+
+  // Number of build rows.
+  int64_t buildSize{};
+
+  // Title for reporting
+  std::string title;
+
+  // This parameter controls the hashing mode. It is incorporated into the keys
+  // on the build side. If the expected mode is an array, its value is 0. If
+  // the expected mode is a normalized key, its value is 'kArrayHashMaxSize' +
+  // 100 to make the key range > 'kArrayHashMaxSize'. If the expected mode is a
+  // hash, its value is the maximum value of int64_t minus 1 to make the key
+  // range  == 'kRangeTooLarge'.
+  int64_t extraValue{};
+
+  // Number of fields.
+  uint32_t numFields{};
+};
+
+class HashJoinGroupProbeBenchmark : public VectorTestBase {
```

Comment:
What's the reason we are not putting this inside the existing `HashTableBenchmark.cpp`?
Also `s/HashJoinGroupProbeBenchmark/HashGroupProbeBenchmark/`

Reply:
Each Benchmark tests different functions, and some of the test details are different. The code is divided into multiple files to make the code clearer. However, there is indeed a lot of code duplication between these tests. We can consider extracting a common parent class.